Tracking issue for TryFrom/TryInto traits

Tracking issue for https://github.com/rust-lang/rfcs/pull/1542

----

To do:

* [ ] Is existing documentation satisfactory?
* [x] https://github.com/rust-lang/rust/pull/56796 Change bounds on `TryFrom` blanket impl to use `Into` instead of `From`
* [ ] (Re)stabilization PR

I haven't read the other comments, but `TryFrom` is broken for me right now (it worked fine before).

rustc version: 

```
rustc 1.22.0-nightly (d6d711dd8 2017-10-10)
binary: rustc
commit-hash: d6d711dd8f7ad5885294b8e1f0009a23dc1f8b1f
commit-date: 2017-10-10
host: x86_64-unknown-linux-gnu
release: 1.22.0-nightly
LLVM version: 4.0
```

The relevant code section that rust complains about is here: https://github.com/fschutt/printpdf/blob/master/src/types/plugins/graphics/two_dimensional/image.rs#L29-L39 and https://github.com/fschutt/printpdf/blob/master/src/types/plugins/graphics/xobject.rs#L170-L200 - it compiled fine a few weeks ago, which is why the library still has the "build passing" badge.

However, on the latest nightly build, `TryFrom` seems to break:

```rust
error[E0119]: conflicting implementations of trait `std::convert::TryFrom<_>` for type `types::plugins::graphics::two_dimensional::image::Image`:
  --> src/types/plugins/graphics/two_dimensional/image.rs:29:1
   |
29 | / impl<T: ImageDecoder> TryFrom<T> for Image {
30 | |     type Error = image::ImageError;
31 | |     fn try_from(image: T)
32 | |     -> std::result::Result<Self, Self::Error>
...  |
38 | |     }
39 | | }
   | |_^
   |
   = note: conflicting implementation in crate `core`

error[E0119]: conflicting implementations of trait `std::convert::TryFrom<_>` for type `types::plugins::graphics::xobject::ImageXObject`:
   --> src/types/plugins/graphics/xobject.rs:170:1
    |
170 | / impl<T: image::ImageDecoder> TryFrom<T> for ImageXObject {
171 | |     type Error = image::ImageError;
172 | |     fn try_from(mut image: T)
173 | |     -> std::result::Result<Self, Self::Error>
...   |
199 | |     }
200 | | }
    | |_^
    |
    = note: conflicting implementation in crate `core`

error: aborting due to 2 previous errors

error: Could not compile `printpdf`.
```

So, supposedly it has a duplicated implementation in crate `core`. If anyone could look into this, that'd be great, thanks.

Hmm yea there is... is something not working properly?

```
error[E0599]: no method named `unwrap` found for type `std::result::Result<usize, <T as std::convert::TryInto<usize>>::Error>` in the current scope
  --> src\types\b8_memory_mapper.rs:67:51
   |
67 |         let address: usize = T::try_into(address).unwrap();
   |                                                   ^^^^^^
   |
   = note: the method `unwrap` exists but the following trait bounds were not satisfied:
           `<T as std::convert::TryInto<usize>>::Error : std::fmt::Debug`
```

--- a/src/libcore/convert.rs
+++ b/src/libcore/convert.rs
@@ -413,6 +413,9 @@ pub trait TryInto<T>: Sized {
 /// As described, [`i32`] implements `TryFrom<i64>`:
 ///
 /// ```
+/// #![feature(try_from)]
+/// use std::convert::TryFrom;
+///
 /// let big_number = 1_000_000_000_000i64;
 /// // Silently truncates `big_number`, requires detecting
 /// // and handling the truncation after the fact.

--- a/src/libcore/convert.rs
+++ b/src/libcore/convert.rs
@@ -415,7 +415,6 @@ pub trait TryInto<T>: Sized {
 /// As described, [`i32`] implements `TryFrom<i64>`:
 ///
 /// ```
-/// #![feature(try_from)]
 /// use std::convert::TryFrom;
 ///
 /// let big_number = 1_000_000_000_000i64;

--- a/src/libcore/convert.rs
+++ b/src/libcore/convert.rs
@@ -366,9 +366,9 @@ pub trait From<T>: Sized {
 /// provides an equivalent `TryInto` implementation for free, thanks to a
 /// blanket implementation in the standard library. For more information on this,
 /// see the documentation for [`Into`].
-/// 
+///
 /// # Implementing `TryInto`
-/// 
+///
 /// This suffers the same restrictions and reasoning as implementing
 /// [`Into`], see there for details.
 ///
@@ -387,25 +387,25 @@ pub trait TryInto<T>: Sized {
 
 /// Simple and safe type conversions that may fail in a controlled
 /// way under some circumstances. It is the reciprocal of [`TryInto`].
-/// 
-/// This is useful when you are doing a type conversion that may 
+///
+/// This is useful when you are doing a type conversion that may
 /// trivially succeed but may also need special handling.
 /// For example, there is no way to convert an `i64` into an `i32`
 /// using the [`From`] trait, because an `i64` may contain a value
 /// that an `i32` cannot represent and so the conversion would lose data.
 /// This might be handled by truncating the `i64` to an `i32` (essentially
-/// giving the `i64`'s value modulo `i32::MAX`) or by simply returning 
-/// `i32::MAX`, or by some other method.  The `From` trait is intended 
-/// for lossless conversions, so the `TryFrom` trait informs the 
+/// giving the `i64`'s value modulo `i32::MAX`) or by simply returning
+/// `i32::MAX`, or by some other method.  The `From` trait is intended
+/// for lossless conversions, so the `TryFrom` trait informs the
 /// programmer when a type conversion could go bad and lets them
 /// decide how to handle it.
-/// 
+///
 /// # Generic Implementations
 ///
 /// - `TryFrom<T> for U` implies [`TryInto<U>`]` for T`
-/// - [`try_from`] is reflexive, which means that `TryFrom<T> for T` 
+/// - [`try_from`] is reflexive, which means that `TryFrom<T> for T`
 /// is implemented
-/// 
+///
 /// # Examples
 ///
 /// As described, [`i32`] implements `TryFrom<i64>`:
@@ -416,10 +416,10 @@ pub trait TryInto<T>: Sized {
 /// // and handling the truncation after the fact.
 /// let smaller_number = big_number as i32;
 /// assert_eq!(smaller_number, -727379968);
-/// 
+///
 /// let try_smaller_number = i32::try_from(big_number);
 /// assert!(try_smaller_number.is_err());
-/// 
+///
 /// let try_successful_smaller_number = i32::try_from(3);
 /// assert!(try_successful_smaller_number.is_ok());
 /// ```

--- a/src/libcore/convert.rs
+++ b/src/libcore/convert.rs
@@ -366,6 +366,11 @@ pub trait From<T>: Sized {
 /// provides an equivalent `TryInto` implementation for free, thanks to a
 /// blanket implementation in the standard library. For more information on this,
 /// see the documentation for [`Into`].
+/// 
+/// # Implementing `TryInto`
+/// 
+/// This suffers the same restrictions and reasoning as implementing
+/// [`Into`], see there for details.
 ///
 /// [`TryFrom`]: trait.TryFrom.html
 /// [`Into`]: trait.Into.html
@@ -380,7 +385,44 @@ pub trait TryInto<T>: Sized {
     fn try_into(self) -> Result<T, Self::Error>;
 }
 
-/// Attempt to construct `Self` via a conversion.
+/// Simple and safe type conversions that may fail in a controlled
+/// way under some circumstances. It is the reciprocal of [`TryInto`].
+/// 
+/// This is useful when you are doing a type conversion that may 
+/// trivially succeed but may also need special handling.
+/// For example, there is no way to convert an `i64` into an `i32`
+/// using the [`From`] trait, because an `i64` may contain a value
+/// that an `i32` cannot represent and so the conversion would lose data.
+/// This might be handled by truncating the `i64` to an `i32` (essentially
+/// giving the `i64`'s value modulo `i32::MAX`) or by simply returning 
+/// `i32::MAX`, or by some other method.  The `From` trait is intended 
+/// for lossless conversions, so the `TryFrom` trait informs the 
+/// programmer when a type conversion could go bad and lets them
+/// decide how to handle it.
+/// 
+/// # Generic Implementations
+///
+/// - `TryFrom<T> for U` implies [`TryInto<U>`]` for T`
+/// - [`try_from`] is reflexive, which means that `TryFrom<T> for T` 
+/// is implemented
+/// 
+/// # Examples
+///
+/// As described, [`i32`] implements `TryFrom<i64>`:
+///
+/// ```
+/// let big_number = 1_000_000_000_000i64;
+/// // Silently truncates `big_number`, requires detecting
+/// // and handling the truncation after the fact.
+/// let smaller_number = big_number as i32;
+/// assert_eq!(smaller_number, -727379968);
+/// 
+/// let try_smaller_number = i32::try_from(big_number);
+/// assert!(try_smaller_number.is_err());
+/// 
+/// let try_successful_smaller_number = i32::try_from(3);
+/// assert!(try_successful_smaller_number.is_ok());
+/// ```
 #[stable(feature = "try_from", since = "1.34.0")]
 pub trait TryFrom<T>: Sized {
     /// The type returned in the event of a conversion error.

--- a/src/libcore/convert.rs
+++ b/src/libcore/convert.rs
@@ -361,11 +361,12 @@ pub trait From<T>: Sized {
 /// An attempted conversion that consumes `self`, which may or may not be
 /// expensive.
 ///
-/// Library authors should not directly implement this trait, but should prefer
-/// implementing the [`TryFrom`] trait, which offers greater flexibility and
-/// provides an equivalent `TryInto` implementation for free, thanks to a
-/// blanket implementation in the standard library. For more information on this,
-/// see the documentation for [`Into`].
+/// Library authors should usually not directly implement this trait,
+/// but should prefer implementing the [`TryFrom`] trait, which offers
+/// greater flexibility and provides an equivalent `TryInto`
+/// implementation for free, thanks to a blanket implementation in the
+/// standard library. For more information on this, see the
+/// documentation for [`Into`].
 ///
 /// # Implementing `TryInto`
 ///
@@ -396,7 +397,7 @@ pub trait TryInto<T>: Sized {
 /// This might be handled by truncating the `i64` to an `i32` (essentially
 /// giving the `i64`'s value modulo `i32::MAX`) or by simply returning
 /// `i32::MAX`, or by some other method.  The `From` trait is intended
-/// for lossless conversions, so the `TryFrom` trait informs the
+/// for perfect conversions, so the `TryFrom` trait informs the
 /// programmer when a type conversion could go bad and lets them
 /// decide how to handle it.
 ///
@@ -404,7 +405,8 @@ pub trait TryInto<T>: Sized {
 ///
 /// - `TryFrom<T> for U` implies [`TryInto<U>`]` for T`
 /// - [`try_from`] is reflexive, which means that `TryFrom<T> for T`
-/// is implemented
+/// is implemented and cannot fail -- the associated `Error` type for
+/// calling `T::try_from()` on a value of type `T` is `!`.
 ///
 /// # Examples
 ///
@@ -417,12 +419,18 @@ pub trait TryInto<T>: Sized {
 /// let smaller_number = big_number as i32;
 /// assert_eq!(smaller_number, -727379968);
 ///
+/// // Returns an error because `big_number` is too big to
+/// // fit in an `i32`.
 /// let try_smaller_number = i32::try_from(big_number);
 /// assert!(try_smaller_number.is_err());
 ///
+/// // Returns `Ok(3)`.
 /// let try_successful_smaller_number = i32::try_from(3);
 /// assert!(try_successful_smaller_number.is_ok());
 /// ```
+///
+/// [`try_from`]: trait.TryFrom.html#tymethod.try_from
+/// [`TryInto`]: trait.TryInto.html
 #[stable(feature = "try_from", since = "1.34.0")]
 pub trait TryFrom<T>: Sized {
     /// The type returned in the event of a conversion error.

--- a/src/libcore/convert.rs
+++ b/src/libcore/convert.rs
@@ -406,7 +406,9 @@ pub trait TryInto<T>: Sized {
 /// - `TryFrom<T> for U` implies [`TryInto<U>`]` for T`
 /// - [`try_from`] is reflexive, which means that `TryFrom<T> for T`
 /// is implemented and cannot fail -- the associated `Error` type for
-/// calling `T::try_from()` on a value of type `T` is `!`.
+/// calling `T::try_from()` on a value of type `T` is `Infallible`.
+/// When the `!` type is stablized `Infallible` and `!` will be
+/// equivalent.
 ///
 /// # Examples
 ///

--- a/src/libcore/num/mod.rs
+++ b/src/libcore/num/mod.rs
@@ -4544,9 +4544,14 @@ macro_rules! try_from_unbounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
-            /// Try to create the target type from the source type.
-            /// This particular variant will never fail, but is included
-            /// for completeness's sake.
+            /// Try to create the target number type from a source
+            /// number type.  If the source type has a larger range
+            /// than the target, or their ranges are disjoint (such
+            /// as converting a signed to unsigned number or vice 
+            /// versa), this will return `None` if the source value
+            /// doesn't fit into the range of the destination value.
+            /// If the conversion can never fail, this is still
+            /// implemented for completeness's sake.
             #[inline]
             fn try_from(value: $source) -> Result<Self, Self::Error> {
                 Ok(value as $target)
@@ -4562,10 +4567,14 @@ macro_rules! try_from_lower_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
-            /// Try to create a target number type from a
-            /// source type that has `source::MIN > dest::MIN`.
-            /// Will return an error if `source` is less than
-            /// `dest::MIN`.
+            /// Try to create the target number type from a source
+            /// number type.  If the source type has a larger range
+            /// than the target, or their ranges are disjoint (such
+            /// as converting a signed to unsigned number or vice 
+            /// versa), this will return `None` if the source value
+            /// doesn't fit into the range of the destination value.
+            /// If the conversion can never fail, this is still
+            /// implemented for completeness's sake.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u >= 0 {
@@ -4585,10 +4594,14 @@ macro_rules! try_from_upper_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
-            /// Try to create a target number type from a
-            /// source type that has `source::MAX > dest::MAX`.
-            /// Will return an error if `source` is greater than
-            /// `dest::MAX`.
+            /// Try to create the target number type from a source
+            /// number type.  If the source type has a larger range
+            /// than the target, or their ranges are disjoint (such
+            /// as converting a signed to unsigned number or vice 
+            /// versa), this will return `None` if the source value
+            /// doesn't fit into the range of the destination value.
+            /// If the conversion can never fail, this is still
+            /// implemented for completeness's sake.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u > (<$target>::max_value() as $source) {
@@ -4608,11 +4621,14 @@ macro_rules! try_from_both_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
-            /// Try to "narrow" a number from the source type
-            /// to the target type.  Will return an error if
-            /// the source value is either larger than the
-            /// `MAX` value for the target type or smaller
-            /// than the `MIN` value for it.
+            /// Try to create the target number type from a source
+            /// number type.  If the source type has a larger range
+            /// than the target, or their ranges are disjoint (such
+            /// as converting a signed to unsigned number or vice 
+            /// versa), this will return `None` if the source value
+            /// doesn't fit into the range of the destination value.
+            /// If the conversion can never fail, this is still
+            /// implemented for completeness's sake.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 let min = <$target>::min_value() as $source;

--- a/src/libcore/num/mod.rs
+++ b/src/libcore/num/mod.rs
@@ -4544,6 +4544,9 @@ macro_rules! try_from_unbounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create the target type from the source type.
+            /// This particular variant will never fail, but is included
+            /// for completeness's sake.
             #[inline]
             fn try_from(value: $source) -> Result<Self, Self::Error> {
                 Ok(value as $target)
@@ -4559,6 +4562,10 @@ macro_rules! try_from_lower_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create a target number type from a
+            /// source type that has `source::MIN > dest::MIN`.
+            /// Will return an error if `source` is less than 
+            /// `dest::MIN`.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u >= 0 {
@@ -4578,6 +4585,10 @@ macro_rules! try_from_upper_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create a target number type from a
+            /// source type that has `source::MAX > dest::MAX`.
+            /// Will return an error if `source` is greater than 
+            /// `dest::MAX`.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u > (<$target>::max_value() as $source) {
@@ -4597,6 +4608,11 @@ macro_rules! try_from_both_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to "narrow" a number from the source type
+            /// to the target type.  Will return an error if 
+            /// the source value is either larger than the 
+            /// `MAX` value for the target type or smaller 
+            /// than the `MIN` value for it.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 let min = <$target>::min_value() as $source;

--- a/src/libcore/num/mod.rs
+++ b/src/libcore/num/mod.rs
@@ -4545,13 +4545,8 @@ macro_rules! try_from_unbounded {
             type Error = TryFromIntError;
 
             /// Try to create the target number type from a source
-            /// number type.  If the source type has a larger range
-            /// than the target, or their ranges are disjoint (such
-            /// as converting a signed to unsigned number or vice 
-            /// versa), this will return `None` if the source value
-            /// doesn't fit into the range of the destination value.
-            /// If the conversion can never fail, this is still
-            /// implemented for completeness's sake.
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(value: $source) -> Result<Self, Self::Error> {
                 Ok(value as $target)
@@ -4568,13 +4563,8 @@ macro_rules! try_from_lower_bounded {
             type Error = TryFromIntError;
 
             /// Try to create the target number type from a source
-            /// number type.  If the source type has a larger range
-            /// than the target, or their ranges are disjoint (such
-            /// as converting a signed to unsigned number or vice 
-            /// versa), this will return `None` if the source value
-            /// doesn't fit into the range of the destination value.
-            /// If the conversion can never fail, this is still
-            /// implemented for completeness's sake.
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u >= 0 {
@@ -4595,13 +4585,8 @@ macro_rules! try_from_upper_bounded {
             type Error = TryFromIntError;
 
             /// Try to create the target number type from a source
-            /// number type.  If the source type has a larger range
-            /// than the target, or their ranges are disjoint (such
-            /// as converting a signed to unsigned number or vice 
-            /// versa), this will return `None` if the source value
-            /// doesn't fit into the range of the destination value.
-            /// If the conversion can never fail, this is still
-            /// implemented for completeness's sake.
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u > (<$target>::max_value() as $source) {
@@ -4622,13 +4607,8 @@ macro_rules! try_from_both_bounded {
             type Error = TryFromIntError;
 
             /// Try to create the target number type from a source
-            /// number type.  If the source type has a larger range
-            /// than the target, or their ranges are disjoint (such
-            /// as converting a signed to unsigned number or vice 
-            /// versa), this will return `None` if the source value
-            /// doesn't fit into the range of the destination value.
-            /// If the conversion can never fail, this is still
-            /// implemented for completeness's sake.
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 let min = <$target>::min_value() as $source;

--- a/src/libcore/num/mod.rs
+++ b/src/libcore/num/mod.rs
@@ -4564,7 +4564,7 @@ macro_rules! try_from_lower_bounded {
 
             /// Try to create a target number type from a
             /// source type that has `source::MIN > dest::MIN`.
-            /// Will return an error if `source` is less than 
+            /// Will return an error if `source` is less than
             /// `dest::MIN`.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
@@ -4587,7 +4587,7 @@ macro_rules! try_from_upper_bounded {
 
             /// Try to create a target number type from a
             /// source type that has `source::MAX > dest::MAX`.
-            /// Will return an error if `source` is greater than 
+            /// Will return an error if `source` is greater than
             /// `dest::MAX`.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
@@ -4609,9 +4609,9 @@ macro_rules! try_from_both_bounded {
             type Error = TryFromIntError;
 
             /// Try to "narrow" a number from the source type
-            /// to the target type.  Will return an error if 
-            /// the source value is either larger than the 
-            /// `MAX` value for the target type or smaller 
+            /// to the target type.  Will return an error if
+            /// the source value is either larger than the
+            /// `MAX` value for the target type or smaller
             /// than the `MIN` value for it.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {

